File: k2/python/k2/properties.py
# Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)

# See ../../../LICENSE for clarification regarding multiple authors

import torch
from torch.utils.dlpack import to_dlpack

from .fsa import Fsa
from _k2 import _is_valid
from _k2 import _is_top_sorted
from _k2 import _is_arc_sorted
from _k2 import _has_self_loops
from _k2 import _is_acyclic
from _k2 import _is_deterministic
from _k2 import _is_epsilon_free
from _k2 import _is_connected
from _k2 import _is_empty


def is_valid(fsa: Fsa) -> bool:
    return _is_valid(fsa.get_base())


def is_top_sorted(fsa: Fsa) -> bool:
    return _is_top_sorted(fsa.get_base())


def is_arc_sorted(fsa: Fsa) -> bool:
    return _is_arc_sorted(fsa.get_base())


def has_self_loops(fsa: Fsa) -> bool:
    return _has_self_loops(fsa.get_base())


def is_acyclic(fsa: Fsa) -> bool:
    return _is_acyclic(fsa.get_base())


def is_deterministic(fsa: Fsa) -> bool:
    return _is_deterministic(fsa.get_base())


def is_epsilon_free(fsa: Fsa) -> bool:
    return _is_epsilon_free(fsa.get_base())


def is_connected(fsa: Fsa) -> bool:
    return _is_connected(fsa.get_base())


def is_empty(fsa: Fsa) -> bool:
    return _is_empty(fsa.get_base())

File: k2/python/host/csrc/properties.cc
// k2/python/csrc/properties.cc

// Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)

// See ../../../LICENSE for clarification regarding multiple authors

#include "k2/python/host/csrc/properties.h"

#include <vector>

#include "k2/csrc/host/array.h"
#include "k2/csrc/host/fsa.h"
#include "k2/csrc/host/properties.h"
#include "k2/python/host/csrc/array.h"

// We would never pass `order` parameter to k2host::IsAcyclic in Python code.
// We can make it accept `None` with `std::optional` in pybind11, but
// that will require C++17, so we here choose to write a version without
// `order`.
static bool IsAcyclic(const k2host::Fsa &fsa) {
  return k2host::IsAcyclic(fsa /*, std::vector<int32_t>* order = nullptr*/);
}

void PybindProperties(py::module &m) {
  m.def("_is_valid", &k2host::IsValid, py::arg("fsa"));
  m.def("_is_top_sorted", &k2host::IsTopSorted, py::arg("fsa"));
  m.def("_is_arc_sorted", &k2host::IsArcSorted, py::arg("fsa"));
  m.def("_has_self_loops", &k2host::HasSelfLoops, py::arg("fsa"));
  m.def("_is_acyclic", &IsAcyclic, py::arg("fsa"));
  m.def("_is_deterministic", &k2host::IsDeterministic, py::arg("fsa"));
  m.def("_is_epsilon_free", &k2host::IsEpsilonFree, py::arg("fsa"));
  m.def("_is_connected", &k2host::IsConnected, py::arg("fsa"));
  m.def("_is_empty", &k2host::IsEmpty, py::arg("fsa"));
}
